### ISSUE TITLE
macos: make libuv tsan-clean

### ISSUE THREAD

**Issue by bnoordhuis** [libuv/libuv#3880]
Follow-up to #3681. ThreadSanitizer is not pleased with ibuv's fsevents glue code and our workaround for kqueue-incompatile file descriptors.

The data races are probably benign but need closer investigation (and fixes.)

Once fixed, consider adding a macos+tsan buildbot.

**Comment by bnoordhuis** [comment]
Not exclusively, I think. See below.
<details>

```
not ok 50 - fs_event_getpath
# exit code 6
# Output from process `fs_event_getpath`:
# ==================
# WARNING: ThreadSanitizer: data race (pid=10412)
#   Write of size 8 at 0x000103e5f428 by main thread:
#     #0 uv_fs_event_init kqueue.c:490 (uv_run_tests_a:x86_64+0x10016659d)
#     #1 run_test_fs_event_getpath test-fs-event.c:1037 (uv_run_tests_a:x86_64+0x100022794)
#     #2 run_test_part runner.c:372 (uv_run_tests_a:x86_64+0x100008c0a)
#     #3 main run-tests.c:81 (uv_run_tests_a:x86_64+0x100005d82)
# 
#   Previous read of size 8 at 0x000103e5f428 by thread T1:
#     #0 uv__fsevents_reschedule fsevents.c:420 (uv_run_tests_a:x86_64+0x10016c0a2)
#     #1 uv__cf_loop_cb fsevents.c:766 (uv_run_tests_a:x86_64+0x10016bdfa)
#     #2 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ <null> (CoreFoundation:x86_64+0x579fa)
# 
#   Location is global 'fs_event' at 0x000103e5f420 (uv_run_tests_a+0x0001002a6428)
# 
#   Thread T1 (tid=77812994, running) created by main thread at:
#     #0 pthread_create <null> (libclang_rt.tsan_osx_dynamic.dylib:x86_64+0x28a3d)
#     #1 uv__fsevents_loop_init fsevents.c:648 (uv_run_tests_a:x
# 86_64+0x100169f37)
#     #2 uv__fsevents_init fsevents.c:806 (uv_run_tests_a:x86_64+0x1001691f8)
#     #3 uv_fs_event_start kqueue.c:538 (uv_run_tests_a:x86_64+0x100166b07)
#     #4 run_test_fs_event_getpath test-fs-event.c:1042 (uv_run_tests_a:x86_64+0x1000228c4)
#     #5 run_test_part runner.c:372 (uv_run_tests_a:x86_64+0x100008c0a)
#     #6 main run-tests.c:81 (uv_run_tests_a:x86_64+0x100005d82)
# 
# SUMMARY: ThreadSanitizer: data race kqueue.c:490 in uv_fs_event_init
# ==================
# ThreadSanitizer: reported 1 warnings
not ok 176 - osx_select
# exit code 6
# Output from process `osx_select`:
# ==================
# WARNING: ThreadSanitizer: data race (pid=10552)
#   Read of size 4 at 0x7ffee7eb7270 by thread T1:
#     #0 uv__io_active core.c:955 (uv_run_tests_a:x86_64+0x100128d4f)
#     #1 uv__stream_osx_select stream.c:163 (uv_run_tests_a:x86_64+0x100150af7)
# 
#   Previous write of size 4 at 0x7ffee7eb7270 by main thread:
#     #0 uv__io_start core.c:883 (uv_run_tests_a:x86_64+0x100127af5)
#     #1 uv__read_start stream.c:1486 (uv_run_tests_a:x86_64+0x100154fd0)
#     #2 uv_read_start uv-common.c:907 (uv_run_tests_a:x86_64+0x100121e8d)
#     #3 run_test_osx_select test-osx-select.c:67 (uv_run_tests_a:x86_64+0x10008485d)
#     #4 run_test_part runner.c:372 (uv_run_tests_a:x86_64+0x100008c0a)
#     #5 main run-tests.c:81 (uv_run_tests_a:x86_64+0x100005d82)
# 
#   Location is stack of main thread.
# 
#   Thread T1 (tid=77814434, running) created by main thread at:
#     #0 pthread_create <null>:8926880 (libclang_rt.tsan_osx_dynamic.dylib:x86_64+0x28a3d)
#     #1 uv_thread_create_ex thread.c:290 (uv_run_tests_a:x86_64+0x10015ac97)
#     #
# 2 uv_thread_create thread.c:244 (uv_run_tests_a:x86_64+0x10015aa93)
#     #3 uv__stream_try_select stream.c:359 (uv_run_tests_a:x86_64+0x100150384)
#     #4 uv_tty_init tty.c:220 (uv_run_tests_a:x86_64+0x10015bcda)
#     #5 run_test_osx_select test-osx-select.c:64 (uv_run_tests_a:x86_64+0x1000847d6)
#     #6 run_test_part runner.c:372 (uv_run_tests_a:x86_64+0x100008c0a)
#     #7 main run-tests.c:81 (uv_run_tests_a:x86_64+0x100005d82)
# 
# SUMMARY: ThreadSanitizer: data race core.c:955 in uv__io_active
# ==================
# got data 1
# ==================
# WARNING: ThreadSanitizer: data race (pid=10552)
#   Read of size 4 at 0x7b3400000098 by thread T1:
#     #0 uv__stream_osx_select stream.c:218 (uv_run_tests_a:x86_64+0x1001512e9)
# 
#   Previous write of size 4 at 0x7b3400000098 by main thread:
#     #0 uv__stream_osx_select_cb stream.c:234 (uv_run_tests_a:x86_64+0x1001505d4)
#     #1 uv__async_io async.c:176 (uv_run_tests_a:x86_64+0x100124f8a)
#     #2 uv__io_poll kqueue.c:371 (uv_run_tests_a:x86_64+0x10016570e)
#     #3 uv_run core.c:410 (uv
# _run_tests_a:x86_64+0x100125d0d)
#     #4 run_test_osx_select test-osx-select.c:78 (uv_run_tests_a:x86_64+0x10008495d)
#     #5 run_test_part runner.c:372 (uv_run_tests_a:x86_64+0x100008c0a)
#     #6 main run-tests.c:81 (uv_run_tests_a:x86_64+0x100005d82)
# 
#   Location is heap block of size 208 at 0x7b3400000000 allocated by main thread:
#     #0 malloc <null>:8926896 (libclang_rt.tsan_osx_dynamic.dylib:x86_64+0x4b16a)
#     #1 uv__malloc uv-common.c:77 (uv_run_tests_a:x86_64+0x100117866)
#     #2 uv__stream_try_select stream.c:323 (uv_run_tests_a:x86_64+0x10014fc6b)
#     #3 uv_tty_init tty.c:220 (uv_run_tests_a:x86_64+0x10015bcda)
#     #4 run_test_osx_select test-osx-select.c:64 (uv_run_tests_a:x86_64+0x1000847d6)
#     #5 run_test_part runner.c:372 (uv_run_tests_a:x86_64+0x100008c0a)
#     #6 main run-tests.c:81 (uv_run_tests_a:x86_64+0x100005d82)
# 
#   Thread T1 (tid=77814434, running) created by main thread at:
#     #0 pthread_create <null>:8926944 (libclang_rt.tsan_osx_dynamic.dylib:x86_64+0x28a3d)
#     #1 uv_thread_create_ex th
# read.c:290 (uv_run_tests_a:x86_64+0x10015ac97)
#     #2 uv_thread_create thread.c:244 (uv_run_tests_a:x86_64+0x10015aa93)
#     #3 uv__stream_try_select stream.c:359 (uv_run_tests_a:x86_64+0x100150384)
#     #4 uv_tty_init tty.c:220 (uv_run_tests_a:x86_64+0x10015bcda)
#     #5 run_test_osx_select test-osx-select.c:64 (uv_run_tests_a:x86_64+0x1000847d6)
#     #6 run_test_part runner.c:372 (uv_run_tests_a:x86_64+0x100008c0a)
#     #7 main run-tests.c:81 (uv_run_tests_a:x86_64+0x100005d82)
# 
# SUMMARY: ThreadSanitizer: data race stream.c:218 in uv__stream_osx_select
# ==================
# got data 2
# got data 3
# ThreadSanitizer: reported 2 warnings
not ok 177 - osx_select_many_fds
# exit code 6
# Output from process `osx_select_many_fds`:
# ==================
# WARNING: ThreadSanitizer: data race (pid=10555)
#   Read of size 4 at 0x7ffee02e0250 by thread T1:
#     #0 uv__io_active core.c:955 (uv_run_tests_a:x86_64+0x100128d4f)
#     #1 uv__stream_osx_select stream.c:163 (uv_run_tests_a:x86_64+0x100150af7)
# 
#   Previous write of size 4 at 0x7ffee02e0250 by main thread:
#     #0 uv__io_start core.c:883 (uv_run_tests_a:x86_64+0x100127af5)
#     #1 uv__read_start stream.c:1486 (uv_run_tests_a:x86_64+0x100154fd0)
#     #2 uv_read_start uv-common.c:907 (uv_run_tests_a:x86_64+0x100121e8d)
#     #3 run_test_osx_select_many_fds test-osx-select.c:120 (uv_run_tests_a:x86_64+0x10008508c)
#     #4 run_test_part runner.c:372 (uv_run_tests_a:x86_64+0x100008c0a)
#     #5 main run-tests.c:81 (uv_run_tests_a:x86_64+0x100005d82)
# 
#   Location is stack of main thread.
# 
#   Thread T1 (tid=77814562, running) created by main thread at:
#     #0 pthread_create <null>:8926880 (libclang_rt.tsan_osx_dynamic.dylib:x86_64+0x28a3d)
#     #1 uv_thread_create_ex thread.c:290 (uv_run_tests_a:x86_64+0x10015a
# c97)
#     #2 uv_thread_create thread.c:244 (uv_run_tests_a:x86_64+0x10015aa93)
#     #3 uv__stream_try_select stream.c:359 (uv_run_tests_a:x86_64+0x100150384)
#     #4 uv_tty_init tty.c:220 (uv_run_tests_a:x86_64+0x10015bcda)
#     #5 run_test_osx_select_many_fds test-osx-select.c:117 (uv_run_tests_a:x86_64+0x100085005)
#     #6 run_test_part runner.c:372 (uv_run_tests_a:x86_64+0x100008c0a)
#     #7 main run-tests.c:81 (uv_run_tests_a:x86_64+0x100005d82)
# 
# SUMMARY: ThreadSanitizer: data race core.c:955 in uv__io_active
# ==================
# got data 1
# ==================
# WARNING: ThreadSanitizer: data race (pid=10555)
#   Read of size 4 at 0x7b5400000318 by thread T1:
#     #0 uv__stream_osx_select stream.c:218 (uv_run_tests_a:x86_64+0x1001512e9)
# 
#   Previous write of size 4 at 0x7b5400000318 by main thread:
#     #0 uv__stream_osx_select_cb stream.c:234 (uv_run_tests_a:x86_64+0x1001505d4)
#     #1 uv__async_io async.c:176 (uv_run_tests_a:x86_64+0x100124f8a)
#     #2 uv__io_poll kqueue.c:371 (uv_run_tests_a:x86_64+0x10016570e)
#     #3 u
# v_run core.c:410 (uv_run_tests_a:x86_64+0x100125d0d)
#     #4 run_test_osx_select_many_fds test-osx-select.c:132 (uv_run_tests_a:x86_64+0x1000851f0)
#     #5 run_test_part runner.c:372 (uv_run_tests_a:x86_64+0x100008c0a)
#     #6 main run-tests.c:81 (uv_run_tests_a:x86_64+0x100005d82)
# 
#   Location is heap block of size 584 at 0x7b5400000280 allocated by main thread:
#     #0 malloc <null>:8926896 (libclang_rt.tsan_osx_dynamic.dylib:x86_64+0x4b16a)
#     #1 uv__malloc uv-common.c:77 (uv_run_tests_a:x86_64+0x100117866)
#     #2 uv__stream_try_select stream.c:323 (uv_run_tests_a:x86_64+0x10014fc6b)
#     #3 uv_tty_init tty.c:220 (uv_run_tests_a:x86_64+0x10015bcda)
#     #4 run_test_osx_select_many_fds test-osx-select.c:117 (uv_run_tests_a:x86_64+0x100085005)
#     #5 run_test_part runner.c:372 (uv_run_tests_a:x86_64+0x100008c0a)
#     #6 main run-tests.c:81 (uv_run_tests_a:x86_64+0x100005d82)
# 
#   Thread T1 (tid=77814562, running) created by main thread at:
#     #0 pthread_create <null>:8926944 (libclang_rt.tsan_osx_dynamic.dylib:x86_6
# 4+0x28a3d)
#     #1 uv_thread_create_ex thread.c:290 (uv_run_tests_a:x86_64+0x10015ac97)
#     #2 uv_thread_create thread.c:244 (uv_run_tests_a:x86_64+0x10015aa93)
#     #3 uv__stream_try_select stream.c:359 (uv_run_tests_a:x86_64+0x100150384)
#     #4 uv_tty_init tty.c:220 (uv_run_tests_a:x86_64+0x10015bcda)
#     #5 run_test_osx_select_many_fds test-osx-select.c:117 (uv_run_tests_a:x86_64+0x100085005)
#     #6 run_test_part runner.c:372 (uv_run_tests_a:x86_64+0x100008c0a)
#     #7 main run-tests.c:81 (uv_run_tests_a:x86_64+0x100005d82)
# 
# SUMMARY: ThreadSanitizer: data race stream.c:218 in uv__stream_osx_select
# ==================
# got data 2
# got data 3
# ThreadSanitizer: reported 2 warnings
```

</details>

**Comment by bnoordhuis** [comment]
I think this should have been addressed by 56fca44a4.
